File: 300-Longest Increasing Subsequence.py
# dp
class Solution:
    def lengthOfLIS(self, nums: 'List[int]') -> 'int':
        if len(nums) < 2: return len(nums)
        dp = [1] * (len(nums) + 1)
        for i in range(1, len(nums)):
            for j in range(0, i):
                if nums[i] > nums[j]: dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)

File: Algorithms/C++/300-Longest_Increasing_Subsequence.cpp
// Runtime: 40 ms, faster than 45.37% of C++ online submissions for Longest Increasing Subsequence.
// Memory Usage: 8.6 MB, less than 93.75% of C++ online submissions for Longest Increasing Subsequence.
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        if (nums.size() < 2)
            return nums.size();
        vector<int> dp(nums.size() + 1, 1);
        for (int i = 0; i < nums.size(); ++i)
            for (int j = 0; j < i; ++j)
                if (nums[i] > nums[j])
                    dp[i] = max(dp[i], dp[j] + 1);
        return *max_element(dp.begin(), dp.end());
    }
};
